### DISCUSSION TITLE
make third_party issue

### DISCUSSION THREAD

**Discussion by C8504** [google/or-tools#3237] [Build and configuration issues] [answered]
Hi experts,

When I tried to build third_party, failed at cloning protobuf, so could I use local protobuf ? And how to configure? thanks a lot.

[ 12%] Performing download step (git clone) for 'protobuf-populate'
Cloning into 'protobuf-src'...
fatal: The remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
Cloning into 'protobuf-src'...
fatal: unable to access 'https://github.com/protocolbuffers/protobuf.git/': LibreSSL SSL_read: error:02FFF03C:system library:func(4095):Operation timed out, errno 60

**Comment by lperron** [comment, marked as answer]
Clearly not a or-tools issue.

**Reply by C8504** [reply]
I agree with you that is NOT a ortools issue, while make third_party is a deps of make ortools.
So I want to use local protobuf instead of building protobuf. 
And I saw in Makefile.local file, which guide us to use local protobuf instead. However, it doesn't work.

```make
## REQUIRED DEPENDENCIES ##
# By default they will be automatically built -> nothing to define
# Define UNIX_PROTOBUF_DIR to depend on external Protobuf dynamic library
#   e.g. UNIX_PROTOBUF_DIR = /opt/protobuf-x.y.z
# Define UNIX_PROTOC_BINARY to use a custom version.
#   e.g. UNIX_PROTOC_BINARY = /opt/protoc-x.y.z/bin/protoc
#   (default: UNIX_PROTOBUF_DIR/bin/protoc)
```

**Comment by Mizux** [comment]
Using the master branch and the CMake based build you could try:
```sh
cmake -S. -Bbuild
```

aka without `-DBUILD_DEPS=ON` or-tools will try to build using the locally installed dependencies...

**Reply by C8504** [reply]
thanks Mizux, when I use this command, I found I missed a Zlib, Re2, Protobuf, could I know where could I configure them? In Makefile.local or not?
thanks again

**Reply by Mizux** [reply]
first try to read: https://github.com/google/or-tools/tree/stable/cmake#cmake-options

you can enable build of deps X or Y i.e. `-DBUILD_Y=ON`
otherwise we try to find the system installed deps using some cmake find Module `FindY.cmake` there are located in cmake/
https://github.com/google/or-tools/tree/master/cmake

According to the CMake's `find_package()`
> New in version 3.12: Search paths specified in the <PackageName>_ROOT CMake variable and the <PackageName>_ROOT environment variable, where <PackageName> is the package to be found.

ref: https://cmake.org/cmake/help/latest/command/find_package.html#id7

So you try to set the env var Protobuf_ROOT Re2_ROOT etc...

On which platform did you work ?

**Reply by C8504** [reply]
THanks Mizux,
I am working on Mac OS version 12.3.1

I tried to set Protobuf_ROOT to my local protobuf by export ProtoBuf_Root = /Users/jc/protobuf-3.19.4
while, I saw make third_party still use git clone protobuf from github....

**Comment by lperron** [comment]
Makefile.local and make third_party have disappeared.
Make build is only a thin layer on top of cmake.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
